### PR TITLE
really check requiredKernelConfig

### PR DESCRIPTION
**Pull request by teto**
###### Motivation for this change
I want the system.requiredKernelConfig to check the final config rather than a structured config that can be ignored during the kernel configuration generation.

I have several follow up PRs. Here is a meta issue with what I intent to accomplish: https://github.com/NixOS/nixpkgs/issues/69014

The lib.kernel.loadConfig function is a translation to nix of https://github.com/NixOS/nixpkgs/blob/e598c4001fda0088506cdc12bf8053705c46676d/pkgs/os-specific/linux/kernel/generate-config.pl#L124.
Hopefully we can replace the perl part with the nix one afterwards.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by teto** [comment]
@fpletz you looked interested :p

**Comment by edolstra** [comment]
I think we have to be careful about moving computations that are currently done in builder (like `generate-config.pl`) to Nix because NixOS evaluation is already very slow.

**Comment by teto** [comment]
at first, I had moved the check to a check-config.pl but I was not sure how to integrate with nixos/modules/system/boot/kernel.nix (I believe it's an IFD situation ? frowned upon) so I translated it to nix.
Maybe we could disable these checks by default ? nixpkgs currently doesn't run the check  I believe because of:
`assertions = if config.boot.kernelPackages.kernel ? features then []`
so that wouldn't change much

**Comment by teto** [comment]
I've added a commit that kinda reverts to the current behavior, i.e., it only runs the check when the kernel is built with no `features` set, aka when built with linuxManualConfig. This way, the extra penalty should only concern few people.

Another aspect (regardless of the implementation) is that I wanted a way to decouple the check from the build. For instance in nixops, I want to check if guest kernels have VIRTIO without rebuilding their kernels.

**Comment by teto** [comment]
still care. For reference, a new project appeared a few days ago to fix kernel configs https://bitbucket.org/easelab/configfix/src/master/

**Comment by kvtb** [comment]
If this PR is not accepted, should `system.requiredKernelConfig` be removed then?
It is used across nixpkgs but has no effect and that is confusing

File: nixos/modules/system/boot/kernel.nix [hunk 1]
```diff
@@ -321,7 +321,27 @@ in
 
     # nixpkgs kernels are assumed to have all required features
     assertions = if config.boot.kernelPackages.kernel ? features then [] else
```
Comment (kvtb) [inline nixos/modules/system/boot/kernel.nix:323]:
```suggestion
    assertions =
```

**Comment by kvtb** [comment]
And, ... I might be wrong, but it looks like IFD in the present form of Nix-implementation (because of `builtins.readFile`), but  bash/perl/whatsoever-implemenation will not be IFD (just like configchecking of webservers are done)

**Pull request closed without merging**